test/metabase/util/queue_test.clj

```diff
@@ -0,0 +1,88 @@
+(ns metabase.util.queue-test
+  (:require
+   [clojure.test :refer [deftest is testing]]
+   [metabase.test :as mt]
+   [metabase.util :as u]
+   [metabase.util.queue :as queue])
+  (:import
+   (java.util Set)
+   (metabase.util.queue DeduplicatingArrayTransferQueue)))
+
+(set! *warn-on-reflection* true)
+
+(defn- simulate-queue! [queue &
+                        {:keys [realtime-threads realtime-events backfill-events]
+                         :or   {realtime-threads 5}}]
+  (let [sent          (atom 0)
+        dropped       (atom 0)
+        skipped       (atom 0)
+        realtime-fn   (fn []
+                        (let [id (rand-int 1000)]
+                          (doseq [e realtime-events]
+                            (case (queue/maybePut! queue {:thread (str "real-" id) :payload e})
+                              true  (swap! sent inc)
+                              false (swap! dropped inc)
+                              nil   (swap! skipped inc)))))
+        background-fn (fn []
+                        (doseq [e backfill-events]
+                          (queue/blockingPut! queue {:thread "back", :payload e})))
+        run!          (fn [f]
+                        (future (f)))]
+
+    (run! background-fn)
+    (future
+     (dotimes [_ realtime-threads]
+       (run! realtime-fn)))
+
+    (let [processed (volatile! [])]
+      (try
+        (while true
+          ;; Stop the consumer once we are sure that there are no more events coming.
+          (u/with-timeout 100
+            (vswap! processed conj (:payload (queue/blockingTake! queue)))
+            ;; Sleep to provide some backpressure
+            (Thread/sleep 1)))
+        (assert false "this is never reached")
+        (catch Exception _
+          {:processed @processed
+           :sent      @sent
+           :dropped   @dropped
+           :skipped   @skipped})))))
+
+(deftest deduplicating-bounded-blocking-queue-test
+  (doseq [dedupe? [true false]]
+    (let [realtime-event-count 500
+          backfill-event-count 1000
+          capacity             (- realtime-event-count 100)
+          ;; Enqueue background events from oldest to newest
+          backfill-events      (range backfill-event-count)
+          ;; Enqueue realtime events from newest to oldest
+          realtime-events      (take realtime-event-count (reverse backfill-events))
+          queue                (queue/bounded-transfer-queue capacity :sleep-ms 10 :block-ms 10 :dedupe? dedupe?)
```

Comment:
why naked keys rather than a map though?

Reply:
I forgot which one we prefer 😄

Happy to update the usage, but think I'll leave the overloading in the signature.

Reply:
Well, I think maps are more honest, they prevent errors (like not supplying value to a key), and also easier to line up nicely 😁

Reply:
How do they prevent forgetting a key, by adding a schema to the argument?

Reply:
They will complain about `{:key}` that it lacks value 😁 like, not much of a safe thingie, just naked values don't look as regular.